gquant/dataframe_flow/node.py

```diff
@@ -284,44 +562,193 @@ def load_cache(self, filename):
+
+    def __delayed_call(self, inputs):
+        def get_pout(out_dict, port):
+            '''Get the output in out_dict at key port. Used for delayed
+            unpacking.'''
+            # DEBUGGING
+            # try:
+            #     from dask.distributed import get_worker
+            #     worker = get_worker()
+            #     print('worker{} get_pout NODE "{}" port "{}" worker: {}'
+            #           .format(worker.name, self.uid, port, worker))
+            # except Exception as err:
+            #     print(err)
+
+            df_out = out_dict.get(port, cudf.DataFrame())
+
+            if isinstance(df_out, cudf.DataFrame):
+                return df_out.copy(deep=False)
+
+            return df_out
+
+        inputs_dly = {}
+        # A dask_cudf object will return a list of dask delayed object using
+        # to_delayed() API. Below the logic assumes (otherwise error) that
+        # all inputs are dask_cudf objects and are distributed in the same
+        # manner. Ex. inputs_dly:
+        #     inputs_dly = {
+        #         p0: {
+        #             iport0: ddf_dly_i0_p0,
+        #             iport1: ddf_dly_i1_p0,
+        #             ... for all iports
+        #         },
+        #         p1: {
+        #             iport0: ddf_dly_i0_p1,
+        #             iport1: ddf_dly_i1_p1,
+        #             ... for all iports
+        #         },
+        #         ... for all partitions
+        # i_x - iport
+        # p_x - partition index
+
+        npartitions = None
+        for iport, dcudf in inputs.items():
+            ddf_dly_list = dcudf.to_delayed()
+            npartitions_ = len(ddf_dly_list)
+            if npartitions is None:
+                npartitions = npartitions_
+            if npartitions != npartitions_:
+                raise Exception(
+                    'Error DASK_CUDF PARTITIONS MISMATCH: Node "{}" input "{}"'
+                    ' has {} npartitions and other inputs have {} partitions'
+                    .format(self.uid, npartitions, npartitions_))
+            for idly, dly in enumerate(ddf_dly_list):
+                inputs_dly.setdefault(idly, {}).update({
+                    iport: dly.persist()
+                })
+
+        # DEBUGGING
+        # print('INPUTS_DLY:\n{}'.format(inputs_dly))
+
+        outputs_dly = {}
+        # Formulate a list of delayed objects for each output port to be able
+        # to call from_delayed to synthesize a dask_cudf object.
+        # Ex. outputs_dly:
+        #     outputs_dly = {
+        #         o0: [ddf_dly_o0_p0, ddf_dly_o0_p1, ... _pN]
+        #         o1: [ddf_dly_o1_p0, ddf_dly_o1_p1, ... _pN]
+        #         ... for all output ports
+        #     }
+        # o_x - output port
+        # p_x - delayed partition
+
+        # VERY IMPORTANT TO USE PERSIST:
+        # https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.DataFrame.persist
+        # Otherwise process will run several times.
```

Comment:
Not sure why the process will run several times. If the computation is expanded as a bunch of delayed functions. Isn't the Dask automatically merge the same delayed functions together to avoid redundant computations? This DataFrame.persist seems to trigger the computation in the background and save the intermediate results distributedly.

Reply:
You would think so, but no dask did NOT automatically merge the same delayed functions together to avoid redundant computations. I used debug print statements to observe that.

I found that handling the dask_cudf dataframes is fragile in that the distribution was not being done optimally. What was happening was that given a bunch of delayed objects corresponding to npartitions (`ddf_cu.to_delayed()` returns a list of delayed), I then formulate some chain of delayed logic on each partition.
```
out = delayed(self.process)(inputs_for_partition_x)
```
Then I need to populate the output ports.
```
# for each port
oport_out = delayed(get_pout)(out)
```
The dask scheduler is not that smart. If I didn't use persist it would actually call `process` multiple times for each oport. Persist fixed it so that dask scheduler was now informed about where the data resides and `process` would get called only once.

I ended up doing logic equivalent to:
```
# I might have a bug in the current code b/c I persist
# the delayed objects from ddc_cu.to_delayed() prior to
# process. I need to only persist the process.
out = delayed(self.process)(inputs_for_partition_x)
out_per = out.persist()
# for each port
oport_out = delayed(get_pout)(out_per )
```

Also, without persist often a non-optimal worker was chosen by the dask scheduler to extract the port output. Say process happened on worker0 and worker1, but then `get_pout` would always occur on worker0 meaning partition 1 was being move to worker0.

Using persist things seemed to be consistent. I debugged it for a while. But frankly (maybe fixing the bug above will help <https://github.com/avolkov1/gQuant/blob/c29d8e59cfc4b4885f96c36c3cb8711f2a8e5725/gquant/dataframe_flow/node.py#L660>), even with that sometimes the partitions don't get split between workers evenly i.e. partition 0 and partition 1 of dask_cudf will end up on the same worker. With persist at least in this case as well the get_pout will be consistent on what worker it runs and dataframes are not moved unnecessarily between workers.

I didn't observe any negative side-effects to calling persist. It triggers the computation asynchronously and the dask-scheduler is better informed now about where the data resides. From the dask docs:
<https://docs.dask.org/en/latest/dataframe-best-practices.html#persist-intelligently>
>Persist is important because Dask DataFrame is lazy by default. It is a way of telling the cluster that it should start executing the computations that you have defined so far, and that it should try to keep those results in memory. 

Really, if the end user wants a guarantee of optimal distribution of dask_cudf, the only way I was able to achieve that is by using client (client.map for example) directly in the `process` instead of relying on `delayed_process` logic. The optimal distribution is an issue of dask_cudf and dask-cuda scheduler. Dask-cudf and cuda scheduler need to be fixed/improved to somehow specify or enforce dask-cudf partitions to spawn workers as evenly as possible.

Reply:
Okay, it is clear to me now. I am glad that you figure this out. We should let the user know about the limitation of using Dask in gQuant. The distributed nodes can only process on the data within the same partition that it cannot access the other partition data. And the returned dask_cudf data frame for each of the ports have the same partitions.

Reply:
Added comments.
https://github.com/avolkov1/gQuant/blob/d32515c103fd29decdeaf47ca5bb071189c3adb5/gquant/dataframe_flow/_node_flow.py#L568-L575